File: benchmarks/micro/ReadonlyOtherLocalFieldStore.c
#include <stdint.h>
#include <unistd.h>

typedef struct Box {
    int64_t value;
} Box;

__attribute__((noinline))
static int64_t read_value(const Box *box) {
    return box->value;
}

static int64_t accumulate(int64_t seed, int64_t salt) {
    Box left = { seed };
    Box right = { salt };
    left.value = seed + salt;
    int64_t observed = read_value(&right);
    left.value = (seed ^ observed) + 17;
    return left.value + observed;
}

int main(void) {
    int64_t total = (int64_t)getpid();

    for (int32_t i = 0; i < 200000; i += 1) {
        total += accumulate(total % 1000000007, (int64_t)i);
        total %= 1000000007;
    }

    return total <= 0 ? 1 : 0;
}

File: benchmarks/micro/ReadonlyOtherLocalFieldStore.rs
struct BoxValue {
    value: i64,
}

#[inline(never)]
fn read_value(value: &BoxValue) -> i64 {
    value.value
}

fn accumulate(seed: i64, salt: i64) -> i64 {
    let mut left = BoxValue { value: seed };
    let right = BoxValue { value: salt };
    left.value = seed + salt;
    let observed = read_value(&right);
    left.value = (seed ^ observed) + 17;
    left.value + observed
}

fn main() {
    let mut total = std::process::id() as i64;

    for i in 0_i32..200_000 {
        total += accumulate(total % 1_000_000_007, i as i64);
        total %= 1_000_000_007;
    }

    if total <= 0 {
        std::process::exit(1);
    }
}
